mlflow/system_metrics/system_metrics_monitor.py

```diff
@@ -0,0 +1,132 @@
+        self.samples_before_logging = (
+            MLFLOW_SYSTEM_METRICS_SAMPLES_BEFORE_LOGGING.get() or samples_before_logging
+        )
+
+        if isinstance(self.sampling_interval, str):
+            self.sampling_interval = float(self.sampling_interval)
+        if isinstance(self.samples_before_logging, str):
+            self.samples_before_logging = int(self.samples_before_logging)
+
+        self._run_id = run_id
+        self.mlflow_logger = BatchMetricsLogger(self._run_id)
+        self._shutdown_event = threading.Event()
+        self._process = None
+
+    def start(self):
+        """Start monitoring system metrics."""
+        try:
+            self._process = threading.Thread(
+                target=self.monitor,
+                daemon=True,
+                name="SystemMetricsMonitor",
+            )
+            self._process.start()
+            _logger.info("MLflow: started monitoring system metrics.")
+        except Exception as e:
+            _logger.warning(f"MLflow: failed to start monitoring system metrics: {e}")
+            self._process = None
+
+    def monitor(self):
+        """Main monitoring loop, which consistently collect and log system metrics."""
+        from mlflow.tracking.fluent import get_run
+
+        while not self._shutdown_event.is_set():
+            for _ in range(self.samples_before_logging):
+                self.collect_metrics()
+                self._shutdown_event.wait(self.sampling_interval)
+                try:
+                    # Get the MLflow run to check if the run is not RUNNING.
+                    run = get_run(self._run_id)
+                except Exception as e:
+                    _logger.warning(f"MLflow: failed to get mlflow run: {e}.")
+                    return
+                if run.info.status != "RUNNING" or self._shutdown_event.is_set():
+                    # If the mlflow run is terminated or receives the shutdown signal, stop
+                    # monitoring.
+                    return
+            metrics = self.aggregate_metrics()
+            try:
+                self.publish_metrics(metrics)
+            except Exception as e:
+                _logger.warning(
+                    f"MLflow: failed to log system metrics: {e}, this is expected if the "
+                    "experiment/run is already terminated."
+                )
+                return
+
+    def collect_metrics(self):
+        """Collect system metrics."""
+        metrics = {}
+        for monitor in self.monitors:
+            monitor.collect_metrics()
+            metrics.update(monitor._metrics)
+        return metrics
+
+    def aggregate_metrics(self):
+        """Aggregate collected metrics."""
+        metrics = {}
+        for monitor in self.monitors:
+            metrics.update(monitor.aggregate_metrics())
+        return metrics
+
+    def publish_metrics(self, metrics):
```

Comment:
```suggestion
    def log_metrics(self, metrics):
```

I think `log_metrics` is more precise.

Reply:
or

```suggestion
    def flush_metrics(self, metrics):
```

Reply:
I took the name `publish_metrics ` from wandb. `log_metrics` is the best option, but we have `mlflow.log_metrics`, which could be confusing. `flush_metrics` is less ideal than `public_metrics` IMO, as the latter one indicates that these metrics are going to the server.

Reply:
makes sense!